File: src/main/java/com/gestion_actas/gestion/de/actas/services/CSVservice.java
package com.gestion_actas.gestion.de.actas.services;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.gestion_actas.gestion.de.actas.model.Actas_Personal;
import com.gestion_actas.gestion.de.actas.repository.ActasPersonalRepository;
import com.opencsv.CSVParser;
import com.opencsv.CSVParserBuilder;
import com.opencsv.CSVReader;
import com.opencsv.CSVReaderBuilder;
import com.opencsv.exceptions.CsvException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.*;
import java.math.BigDecimal;
import java.nio.charset.StandardCharsets;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.*;

@Service
public class CSVservice {

    @Autowired
    private ActasPersonalRepository repository2;

    @Autowired
    private ObjectMapper objectMapper;

    public List<Actas_Personal> importarCsv(MultipartFile file) throws IOException, CsvException {
        List<Actas_Personal> lista = new ArrayList<>();

        // Leemos bytes para poder "asomar" la primera línea y luego volver a leer desde el principio
        byte[] bytes = file.getBytes();

        // Detecta el separador mirando la primera línea no vacía
        String headerLine = readFirstNonEmptyLine(bytes);
        if (headerLine == null) return lista;
        char separator = detectSeparator(headerLine); // ',' o ';'

        CSVParser parser = new CSVParserBuilder()
                .withSeparator(separator)
                .withIgnoreQuotations(false)
                .build();

        try (CSVReader reader = new CSVReaderBuilder(
                new InputStreamReader(new ByteArrayInputStream(bytes), StandardCharsets.UTF_8))
                .withCSVParser(parser)
                .build()) {

            List<String[]> rows = reader.readAll();
            if (rows.isEmpty()) return lista;

            String[] header = rows.get(0);

            Map<String, Integer> headerMap = new HashMap<>();
            for (int i = 0; i < header.length; i++) {
                headerMap.put(normalizaNombre(header[i]), i);
            }

            int ingresosIdx = headerMap.getOrDefault(normalizaNombre("INGRESOS"), -1);
            int egresosIdx = headerMap.getOrDefault(normalizaNombre("EGRESOS"), -1);
            if (ingresosIdx == -1) ingresosIdx = header.length;
            if (egresosIdx == -1) egresosIdx = header.length;

            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("d/M/yyyy");

            for (int i = 1; i < rows.size(); i++) {
                String[] data = rows.get(i);
                Actas_Personal p = new Actas_Personal();

                p.setPeriodoPago(getValue(data, headerMap, "PeriodoPago"));
                p.setFechaIngreso(parseFecha(getValue(data, headerMap, "FechaIngreso"), formatter));
                p.setFechaTermino(parseFecha(getValue(data, headerMap, "FechaTérmino"), formatter));
                p.setDocReferencia(getValue(data, headerMap, "DocReferencia"));
                p.setCodigoModular(getValue(data, headerMap, "CODIGO MODULAR"));
                p.setAirhsp(getValue(data, headerMap, "AIRHSP"));
                p.setCargo(getValue(data, headerMap, "CARGO"));
                p.setApePaterno(getValue(data, headerMap, "ApePaterno"));
                p.setApeMaterno(getValue(data, headerMap, "ApeMaterno"));
                p.setNombres(getValue(data, headerMap, "Nombres"));
                p.setSituacion(getValue(data, headerMap, "Situación"));
                p.setDiasLicencia(parseInt(getValue(data, headerMap, "diaslicencia")));
                p.setFecinilic(parseFecha(getValue(data, headerMap, "fecinilic"), formatter));
                p.settPlanilla(getValue(data, headerMap, "T_Planilla"));
                p.setFechaNacimiento(parseFecha(getValue(data, headerMap, "FechaNacimiento"), formatter));
                p.setSexo(getValue(data, headerMap, "Sexo"));
                p.setTipoDocumento(getValue(data, headerMap, "TipoDocumento"));
                p.setNroDocumento(getValue(data, headerMap, "NroDocumento"));
                p.setIpss(getValue(data, headerMap, "IPSS"));
                p.setRegPensionario(getValue(data, headerMap, "RegPensionario"));
                p.setCodNexus(getValue(data, headerMap, "CodNexus"));
                p.setAfp(getValue(data, headerMap, "AFP"));
                p.setCuspp(getValue(data, headerMap, "CUSPP"));
                p.setFechaAfiliacion(parseFecha(getValue(data, headerMap, "FechaAfiliación"), formatter));
                p.setFechaDevengue(parseFecha(getValue(data, headerMap, "FechaDevengue"), formatter));
                p.setUgel(getValue(data, headerMap, "Ugel"));
                p.setCodEstablecimiento(getValue(data, headerMap, "CodEstablecimiento"));
                p.setEstablecimiento(getValue(data, headerMap, "Establecimiento"));
                p.setNivel(getValue(data, headerMap, "Nivel"));
                p.setCaractEstablecimiento(parseInt(getValue(data, headerMap, "CaractEstablecimiento")));
                p.setUnidCosteo(getValue(data, headerMap, "UnidCosteo"));
                p.setCargoOrig(getValue(data, headerMap, "Cargo/Orig"));
                p.setLeyendaPermanente(getValue(data, headerMap, "LeyendaPermanente"));
                p.setModoPago(getValue(data, headerMap, "ModoPago"));
                p.setCtaCte(getValue(data, headerMap, "CtaCte"));
                p.setDiasTrabajados(parseInt(getValue(data, headerMap, "DíasTrabajados")));
                p.setDecimas(parseInt(getValue(data, headerMap, "Décimas")));
                p.setRegLaboral(getValue(data, headerMap, "RegLaboral"));
                p.setTipoServidor(getValue(data, headerMap, "TipoServidor"));
                p.setNivelMagisterial(getValue(data, headerMap, "NivelMagisterial"));
                p.setCodcomuna(getValue(data, headerMap, "codcomuna"));
                p.setGrupoRemunerativo(getValue(data, headerMap, "GrupoRemunerativo"));
                p.setJornadaLaboral(getValue(data, headerMap, "JornadaLaboral"));
                p.setTiempoServicio(getValue(data, headerMap, "TiempoServicio"));
                p.setCadPresupuestal(getValue(data, headerMap, "CadPresupuestal"));
                p.setTimpopen(getValue(data, headerMap, "timpopen"));
                p.setTributable(parseBigDecimal(getValue(data, headerMap, "tributable")));
                p.setImponible(parseBigDecimal(getValue(data, headerMap, "imponible")));
                p.setTotalRemuneracion(parseBigDecimal(getValue(data, headerMap, "TotalRemuneración")));
                p.setTotalLiquido(parseBigDecimal(getValue(data, headerMap, "TotalLiquido")));

                // INGRESOS
                Map<String, BigDecimal> ingresosMap = new LinkedHashMap<>();
                if (ingresosIdx >= 0 && ingresosIdx < header.length) {
                    int end = (egresosIdx > ingresosIdx) ? egresosIdx : header.length;
                    for (int idx = ingresosIdx + 1; idx < end && idx < header.length && idx < data.length; idx++) {
                        String colName = header[idx] != null ? header[idx].trim() : ("Col_" + idx);
                        ingresosMap.put(colName, parseBigDecimal(safeCell(data, idx)));
                    }
                }
                p.setIngresosJson(objectMapper.writeValueAsString(ingresosMap));

                // EGRESOS
                Map<String, BigDecimal> egresosMap = new LinkedHashMap<>();
                if (egresosIdx >= 0 && egresosIdx < header.length) {
                    for (int idx = egresosIdx + 1; idx < header.length && idx < data.length; idx++) {
                        String colName = header[idx] != null ? header[idx].trim() : ("Col_" + idx);
                        egresosMap.put(colName, parseBigDecimal(safeCell(data, idx)));
                    }
                }
                p.setEgresosJson(objectMapper.writeValueAsString(egresosMap));

                lista.add(p);
            }
        }

        repository2.saveAll(lista);
        return lista;
    }

    private String readFirstNonEmptyLine(byte[] bytes) throws IOException {
        try (BufferedReader br = new BufferedReader(new InputStreamReader(new ByteArrayInputStream(bytes), StandardCharsets.UTF_8))) {
            String line;
            while ((line = br.readLine()) != null) {
                if (line.startsWith("\uFEFF")) line = line.substring(1); // quita BOM
                if (!line.trim().isEmpty()) return line;
            }
            return null;
        }
    }

    private char detectSeparator(String headerLine) {
        int commas = countChar(headerLine, ',');
        int semis = countChar(headerLine, ';');
        // Si aparecen ambos, elegimos el más frecuente
        if (semis > commas) return ';';
        // por defecto coma
        return ',';
    }

    private int countChar(String s, char c) {
        int n = 0;
        for (int i = 0; i < s.length(); i++) if (s.charAt(i) == c) n++;
        return n;
    }

    private String safeCell(String[] data, int idx) {
        return (idx >= 0 && idx < data.length) ? data[idx] : null;
    }

    private String normalizaNombre(String s) {
        if (s == null) return "";
        s = s.replace("\uFEFF", ""); // BOM
        s = s.replaceAll("[\\p{C}\\p{Z}]", "");
        s = s.replaceAll("\\s+", "");
        s = java.text.Normalizer.normalize(s, java.text.Normalizer.Form.NFD)
                .replaceAll("\\p{InCombiningDiacriticalMarks}+", "");
        s = s.replaceAll("[_\\-\\./]", "");
        s = s.toLowerCase();
        return s;
    }

    private String getValue(String[] data, Map<String,Integer> headerMap, String colName) {
        String key = normalizaNombre(colName);
        Integer idx = headerMap.get(key);
        if (idx != null && idx < data.length) {
            return data[idx];
        }
        return null;
    }

    private LocalDate parseFecha(String val, DateTimeFormatter formatter) {
        if (val == null) return null;
        val = val.trim();
        if (val.isEmpty() || val.equals("1/1/1900") || val.equals("01/01/1900") || val.equals("00/01/1900")) return null;
        try { return LocalDate.parse(val, formatter); }
        catch (Exception e) { return null; }
    }

    private Integer parseInt(String val) {
        try {
            if (val == null) return null;
            val = val.trim();
            return val.isEmpty() ? null : Integer.valueOf(val);
        } catch (Exception e) {
            return null;
        }
    }

    private BigDecimal parseBigDecimal(String val) {
        try {
            if (val == null) return BigDecimal.ZERO;
            String s = val.trim();
            if (s.isEmpty()) return BigDecimal.ZERO;

            if (s.matches("^-?\\d{1,3}(\\.\\d{3})*,\\d+$")) {
                s = s.replace(".", "").replace(",", ".");
            } else if (s.matches("^-?\\d{1,3}(,\\d{3})*\\.\\d+$")) {
                s = s.replace(",", "");
            }
            return new BigDecimal(s);
        } catch (Exception e) {
            return BigDecimal.ZERO;
        }
    }
}

File: BackendActasPHP/app/Services/CSVService.php
<?php

namespace App\Services;

use App\Models\ActasPersonal;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Log;
use Illuminate\Support\Facades\File;

class CSVService
{
    // Importa el CSV y guarda los registros en la base de datos
    public function importarCsv($file)
    {
        $lista = [];

        // Leer el archivo completo en memoria
        $bytes = File::get($file->getRealPath());

        // Detectar la primera línea no vacía y el separador
        $headerLine = $this->readFirstNonEmptyLine($bytes);
        if ($headerLine === null) return $lista;
        $separator = $this->detectSeparator($headerLine);

        // Convertir el archivo en array de líneas
        $lines = preg_split("/(\r\n|\n|\r)/", $bytes);
        $rows = [];
        foreach ($lines as $line) {
            if (strlen(trim($line)) > 0) {
                $rows[] = str_getcsv($line, $separator);
            }
        }
        if (empty($rows)) return $lista;
        $header = $rows[0];

        // Mapear nombres normalizados a índices
        $headerMap = [];
        foreach ($header as $i => $col) {
            $headerMap[$this->normalizaNombre($col)] = $i;
        }

        $ingresosIdx = $headerMap[$this->normalizaNombre('INGRESOS')] ?? count($header);
        $egresosIdx = $headerMap[$this->normalizaNombre('EGRESOS')] ?? count($header);

        $formatter = 'd/m/Y'; // Formato fecha

        for ($i = 1; $i < count($rows); $i++) {
            $data = $rows[$i];
            $p = new ActasPersonal();

            // Asignar los datos generales
            $p->periodoPago = $this->getValue($data, $headerMap, 'PeriodoPago');
            $p->fechaIngreso = $this->parseFecha($this->getValue($data, $headerMap, 'FechaIngreso'), $formatter);
            $p->fechaTermino = $this->parseFecha($this->getValue($data, $headerMap, 'FechaTérmino'), $formatter);
            $p->docReferencia = $this->getValue($data, $headerMap, 'DocReferencia');
            $p->codigoModular = $this->getValue($data, $headerMap, 'CODIGO MODULAR');
            $p->airhsp = $this->getValue($data, $headerMap, 'AIRHSP');
            $p->cargo = $this->getValue($data, $headerMap, 'CARGO');
            $p->apePaterno = $this->getValue($data, $headerMap, 'ApePaterno');
            $p->apeMaterno = $this->getValue($data, $headerMap, 'ApeMaterno');
            $p->nombres = $this->getValue($data, $headerMap, 'Nombres');
            $p->situacion = $this->getValue($data, $headerMap, 'Situación');
            $p->diasLicencia = $this->parseInt($this->getValue($data, $headerMap, 'diaslicencia'));
            $p->fecinilic = $this->parseFecha($this->getValue($data, $headerMap, 'fecinilic'), $formatter);
            $p->tPlanilla = $this->getValue($data, $headerMap, 'T_Planilla');
            $p->fechaNacimiento = $this->parseFecha($this->getValue($data, $headerMap, 'FechaNacimiento'), $formatter);
            $p->sexo = $this->getValue($data, $headerMap, 'Sexo');
            $p->tipoDocumento = $this->getValue($data, $headerMap, 'TipoDocumento');
            $p->nroDocumento = $this->getValue($data, $headerMap, 'NroDocumento');
            $p->ipss = $this->getValue($data, $headerMap, 'IPSS');
            $p->regPensionario = $this->getValue($data, $headerMap, 'RegPensionario');
            $p->codNexus = $this->getValue($data, $headerMap, 'CodNexus');
            $p->afp = $this->getValue($data, $headerMap, 'AFP');
            $p->cuspp = $this->getValue($data, $headerMap, 'CUSPP');
            $p->fechaAfiliacion = $this->parseFecha($this->getValue($data, $headerMap, 'FechaAfiliación'), $formatter);
            $p->fechaDevengue = $this->parseFecha($this->getValue($data, $headerMap, 'FechaDevengue'), $formatter);
            $p->ugel = $this->getValue($data, $headerMap, 'Ugel');
            $p->codEstablecimiento = $this->getValue($data, $headerMap, 'CodEstablecimiento');
            $p->establecimiento = $this->getValue($data, $headerMap, 'Establecimiento');
            $p->nivel = $this->getValue($data, $headerMap, 'Nivel');
            $p->caractEstablecimiento = $this->parseInt($this->getValue($data, $headerMap, 'CaractEstablecimiento'));
            $p->unidCosteo = $this->getValue($data, $headerMap, 'UnidCosteo');
            $p->cargoOrig = $this->getValue($data, $headerMap, 'Cargo/Orig');
            $p->leyendaPermanente = $this->getValue($data, $headerMap, 'LeyendaPermanente');
            $p->modoPago = $this->getValue($data, $headerMap, 'ModoPago');
            $p->ctaCte = $this->getValue($data, $headerMap, 'CtaCte');
            $p->diasTrabajados = $this->parseInt($this->getValue($data, $headerMap, 'DíasTrabajados'));
            $p->decimas = $this->parseInt($this->getValue($data, $headerMap, 'Décimas'));
            $p->regLaboral = $this->getValue($data, $headerMap, 'RegLaboral');
            $p->tipoServidor = $this->getValue($data, $headerMap, 'TipoServidor');
            $p->nivelMagisterial = $this->getValue($data, $headerMap, 'NivelMagisterial');
            $p->codcomuna = $this->getValue($data, $headerMap, 'codcomuna');
            $p->grupoRemunerativo = $this->getValue($data, $headerMap, 'GrupoRemunerativo');
            $p->jornadaLaboral = $this->getValue($data, $headerMap, 'JornadaLaboral');
            $p->tiempoServicio = $this->getValue($data, $headerMap, 'TiempoServicio');
            $p->cadPresupuestal = $this->getValue($data, $headerMap, 'CadPresupuestal');
            $p->timpopen = $this->getValue($data, $headerMap, 'timpopen');
            $p->tributable = $this->parseBigDecimal($this->getValue($data, $headerMap, 'tributable'));
            $p->imponible = $this->parseBigDecimal($this->getValue($data, $headerMap, 'imponible'));
            $p->totalRemuneracion = $this->parseBigDecimal($this->getValue($data, $headerMap, 'TotalRemuneración'));
            $p->totalLiquido = $this->parseBigDecimal($this->getValue($data, $headerMap, 'TotalLiquido'));

            // INGRESOS JSON
            $ingresosMap = [];
            if ($ingresosIdx >= 0 && $ingresosIdx < count($header)) {
                $end = ($egresosIdx > $ingresosIdx) ? $egresosIdx : count($header);
                for ($idx = $ingresosIdx + 1; $idx < $end && $idx < count($header) && $idx < count($data); $idx++) {
                    $colName = $header[$idx] ?? ("Col_" . $idx);
                    $ingresosMap[$colName] = $this->parseBigDecimal($data[$idx] ?? null);
                }
            }
            $p->ingresosJson = json_encode($ingresosMap, JSON_UNESCAPED_UNICODE);

            // EGRESOS JSON
            $egresosMap = [];
            if ($egresosIdx >= 0 && $egresosIdx < count($header)) {
                for ($idx = $egresosIdx + 1; $idx < count($header) && $idx < count($data); $idx++) {
                    $colName = $header[$idx] ?? ("Col_" . $idx);
                    $egresosMap[$colName] = $this->parseBigDecimal($data[$idx] ?? null);
                }
            }
            $p->egresosJson = json_encode($egresosMap, JSON_UNESCAPED_UNICODE);

            $p->save();
            $lista[] = $p;
        }

        return $lista;
    }

    // Helpers

    private function readFirstNonEmptyLine($bytes)
    {
        $lines = preg_split("/(\r\n|\n|\r)/", $bytes);
        foreach ($lines as $line) {
            if (strpos($line, "\uFEFF") === 0) $line = substr($line, 1);
            if (trim($line) !== '') return $line;
        }
        return null;
    }

    private function detectSeparator($headerLine)
    {
        $commas = substr_count($headerLine, ',');
        $semis = substr_count($headerLine, ';');
        return ($semis > $commas) ? ';' : ',';
    }

    private function normalizaNombre($s)
    {
        if ($s === null) return "";
        $s = str_replace("\uFEFF", "", $s); // BOM
        $s = preg_replace('/[\p{C}\p{Z}]/u', '', $s);
        $s = preg_replace('/\s+/', '', $s);
        $s = iconv('UTF-8', 'ASCII//TRANSLIT', $s);
        $s = preg_replace('/[_\-\.\/]/', '', $s);
        $s = strtolower($s);
        return $s;
    }

    private function getValue($data, $headerMap, $colName)
    {
        $key = $this->normalizaNombre($colName);
        $idx = $headerMap[$key] ?? null;
        return ($idx !== null && $idx < count($data)) ? $data[$idx] : null;
    }

    private function parseFecha($val, $format)
    {
        if ($val === null) return null;
        $val = trim($val);
        if ($val === '' || $val === '1/1/1900' || $val === '01/01/1900' || $val === '00/01/1900') return null;
        $d = \DateTime::createFromFormat($format, $val);
        return $d ? $d->format('Y-m-d') : null;
    }

    private function parseInt($val)
    {
        try {
            if ($val === null) return null;
            $val = trim($val);
            return ($val === '') ? null : intval($val);
        } catch (\Exception $e) {
            return null;
        }
    }

    private function parseBigDecimal($val)
    {
        try {
            if ($val === null) return 0;
            $s = trim($val);
            if ($s === '') return 0;
            // Europeo: 1.234,56 -> 1234.56
            if (preg_match('/^-?\d{1,3}(\.\d{3})*,\d+$/', $s)) {
                $s = str_replace('.', '', $s);
                $s = str_replace(',', '.', $s);
            } elseif (preg_match('/^-?\d{1,3}(,\d{3})*\.\d+$/', $s)) {
                $s = str_replace(',', '', $s);
            }
            return floatval($s);
        } catch (\Exception $e) {
            return 0;
        }
    }
}